### PR TITLE
[circle-mpqsolver] Replace 'ErrorApproximator' with 'VISQErrorApproximator'

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR replaces 'ErrorApproximator' with 'VISQErrorApproximator'

This PR intoduces:
1) `total absolute error` and `batch_mode` into `visq` component.
2) it replaces `ErrorApproximator` in `circle-mpqsolver` with `VISQErrorApproximator`
3) introduces `circle-mpqsolver` into packaging.

FYI:
for mobile_net_v1 and 200 images it produces the following output:
```
Full int16 model quantization error 9.84288e-05
Full uint8 model quantization error 0.000275709
Target quantization error 0.000187069
Q16 will be set at input due to 
152628 error at input vs 
50324.6 error at output.
16 : 0.000148376
8 : 0.000175282
4 : 0.000189274
6 : 0.000176954
5 : 0.000181785
4 : 0.000189274
Found the best configuration at 5 depth.
Elapsed Time: 19.533 minutes.
```
Output for CARTOON and 200 images:
```
Full int16 model quantization error 0.0037992
Full uint8 model quantization error 0.0617312
Target quantization error 0.0327652
Q8 will be set at input due to 
1.25213e+06 error at input vs 
2.38788e+06 error at output.
31 : 0.0119042
47 : 0.0154902
55 : 0.0219636
59 : 0.0614759
57 : 0.0448715
56 : 0.0244399
Found the best configuration at 56 depth.
Elapsed Time: 40.15 minutes.

```
Issue: #10393
ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
Please add tests for new codes and maybe for existing codes if they are not tests.
Our SE requires code test coverage to be in some portion.
I'm not confident to add to release package: please provide some examples that I can evaluate.

**Comment by stamalakhov** [comment]
> Please add tests for new codes and maybe for existing codes if they are not tests. Our SE requires code test coverage to be in some portion. I'm not confident to add to release package: please provide some examples that I can evaluate.

Ok.

**Comment by seanshpark** [comment]
@stamalakhov , @jinevening , I don't like this tool, made of C/C++, calling `visq` that is made of python.

**Comment by stamalakhov** [comment]
> @stamalakhov , @jinevening , I don't like this tool, made of C/C++, calling `visq` that is made of python.

@jinevening @seanshpark 
Thank you.
M.b. i can try direct evaluation of `total_absolute_error` using c++ 'CircleInterpreter' with attached callbacks will be enough?

**Comment by jinevening** [comment]
Can you use json file generated by visq instead of calling visq in circle-mpqsolver?

**Comment by stamalakhov** [comment]
> Can you use json file generated by visq instead of calling visq in circle-mpqsolver?

Yes. It's possible. Like an additional parameter to command line.

**Comment by jinevening** [comment]
That will separate the roles of the two modules, making them easier to maintain.

**Comment by stamalakhov** [comment]
> That will separate the roles of the two modules, making them easier to maintain.

Ok. Thank you. I see. I'll rework the draft.

**Comment by stamalakhov** [comment]
@jinevening @seanshpark 
Direct calling of `visq` component is excluded.

File: compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp [hunk 1]
```diff
@@ -172,9 +179,17 @@ std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_pat
...
-        error_at_input_is_larger_than_at_output(nodes_depth, 0.5f * (max_depth + min_depth));
-      break;
+    {
+      auto model = read_module(module_path);
+      if (!_quantizer->quantize(model.get(), "uint8", layer_params))
+      {
+        std::cerr << "ERROR: Failed to quantize model" << std::endl;
+        return nullptr;
+      }
+      int16_front = error_at_input_is_larger_than_at_output(nodes_depth, _visq_data_path,
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp:189]: (optioanl) this method is too long. please make it shorter.

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#include <fstream>
+
+using namespace mpqsolver::bisection;
+
+VISQErrorApproximator::VISQErrorApproximator(const std::string &visq_data_path)
+  : _visq_data_path(visq_data_path)
+{
+  if (!init())
+  {
+    throw std::runtime_error("Error in using visq error approximation");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp:31]: please don't throw in ctor

**Comment by seanshpark** [comment]
please fix CI build failure

File: compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp (hunk 1 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp:189]: > (optional) this method is too long. please make it shorter.

Ok.

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp (hunk 2 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp:31]: > please don't throw in ctor
> make user do the init.

ok.

**Comment by stamalakhov** [comment]
> please fix CI build failure

ok

**Comment by jinevening** [comment]
Hmm.. Can you explain why you used TAE as a metric?

Front part and the rear part may have different number of layers, so TAE will depend on the logic to split a model.

Also, it aggregates all errors, so depends on the size of intermediate FMs.

How about using mean error (divide by number of layers, and divide by the size of tensor)?

And, I'm not sure absolute error is the best choice. Squared error is a more commonly used metric.

**Comment by stamalakhov** [comment]
> Hmm.. Can you explain why you used TAE as a metric?

It worked for me, for MobileNetV1 it produced Q16 at front part (layers near inputs) and for CARTOON it produced Q16 at rear part (layers near outputs). I've made experiments with `MPEIR` metric and `MSE` metric and both of them produced the opposite results: for MobileNetV1 it produced Q16 to be at rear part (layers near outputs).


> Front part and the rear part may have different number of layers, so TAE will depend on the logic to split a model.

TAE is layerwise metric and to my knowledge it does not depend on splitting the model.
TAE stands for `total absolute error` 
```
np.sum(np.abs(fp32_data - fq_data))
```
done layerwise, as was done MSE metric, but not squared and no `mean` was taken.
```
np.square(fp32_data - fq_data).mean()
```


> Also, it aggregates all errors, so depends on the size of intermediate FMs.

Seems like the size of intermediate FMs is somehow relevant for the final error. But it's just an empirical suggestion.

**Comment by stamalakhov** [comment]
> And, I'm not sure absolute error is the best choice. Squared error is a more commonly used metric.

Yes. Seems like squared error is possible too, but IMHO it should be total.

**Comment by jinevening** [comment]
> TAE is layerwise metric and to my knowledge it does not depend on splitting the model.

Ah, I see. So, the layer errors are aggregated in `circle-mpqsolver`, not `visq`.

We may go on with total errors and then test other models later.

**Comment by stamalakhov** [comment]
> Ah, I see. So, the layer errors are aggregated in `circle-mpqsolver`, not `visq`.
> 
> We may go on with total errors and then test other models later.

Ok. Thank you.

**Comment by stamalakhov** [comment]
@seanshpark @jinevening 
Seems like errors related to my features are fixed. When the draft will be splitted into separate PRs failure of "nncc-jammy-debug" will be gone due to #10403.

**Comment by seanshpark** [comment]
plz read https://github.com/Samsung/ONE/pull/10432#issuecomment-1419866198

**Comment by stamalakhov** [comment]
> plz read [#10432 (comment)](https://github.com/Samsung/ONE/pull/10432#issuecomment-1419866198)

Ahh. Sorry. I'll provide tests.

**Comment by stamalakhov** [comment]
@seanshpark 
Tests were added.

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp [hunk 3]
```diff
@@ -32,6 +32,19 @@ void print_version(void)
   std::cout << vconone::get_copyright() << std::endl;
 }
 
+int handleAutoAlgorithm(arser::Arser &arser, mpqsolver::bisection::BisectionSolver &solver)
+{
+  solver.algorithm(mpqsolver::bisection::BisectionSolver::Algorithm::Auto);
+  auto data_path = arser.get<std::string>("--visq_data");
+  if (data_path.empty())
+  {
+    std::cerr << "ERROR: for auto mode you must provide valid visq_file" << std::endl;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:41]: Q) above option checks `visq_data` but message is `visq_file`. is this intended?

Comment (seanshpark) [inline compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:41]: (optional) `please provide visq_file for auto mode` ?

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp [hunk 4]
```diff
@@ -32,6 +32,19 @@ void print_version(void)
   std::cout << vconone::get_copyright() << std::endl;
 }
 
+int handleAutoAlgorithm(arser::Arser &arser, mpqsolver::bisection::BisectionSolver &solver)
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:35]: (optional) `handleAutoAlgorithm` -> `checkAutoAlgorithm` ?

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp [hunk 5]
```diff
@@ -73,6 +86,11 @@ int entry(int argc, char **argv)
     .help("Data type of quantized model's outputs (default: uint8)");
 
   arser.add_argument("--output_model").required(true).help("Output quantized model");
+  arser.add_argument("--visq_data")
+    .type(arser::DataType::STR)
+    .default_value("")
+    .required(false)
+    .help("Output quantized model");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:93]: is this correct help message?

File: compiler/circle-mpqsolver/src/bisection/BisectionSolver.h [hunk 6]
```diff
@@ -64,6 +64,11 @@ class BisectionSolver final : public MPQSolver
    */
   void algorithm(Algorithm algorithm);
 
+  /**
+   * @brief set path of visq_data
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/BisectionSolver.h:68]: this message doesn't seem to provide a help. (1) remove this (2) provide better message.

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+#include <luci/Log.h>
+
+#include <json.h>
+#include <fstream>
+
+using namespace mpqsolver::bisection;
+
+void VISQErrorApproximator::init(const std::string &visq_data_path)
+{
+  LOGGER(l);
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.cpp:28]: doesn't seem to be used

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * limitations under the License.
+ */
+
+#ifndef __MPQSOLVER_BISECTION_VISQ_ERROR_APPROXIMATOR_H__
+#define __MPQSOLVER_BISECTION_VISQ_ERROR_APPROXIMATOR_H__
+
+#include <loco.h>
+
+#include <luci/IR/CircleNodeDecl.h>
+#include <luci/IR/Module.h>
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.h:23]: are these headers really needed here?

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.test.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.test.cpp:2]: 2023?

File: compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.test.cpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include "VISQErrorApproximator.h"
+#include <json.h>
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/VISQErrorApproximator.test.cpp:20]:
```suggestion
#include "VISQErrorApproximator.h"

#include <json.h>
#include <gtest/gtest.h>
```

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp (hunk 3 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:41]: no. i'll change it to visq_file

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp (hunk 4 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:35]: IMHO the function does not checks but also sets some properties for solver, e.g. sets used `algorithm`.

File: compiler/circle-mpqsolver/src/CircleMPQSolver.cpp (hunk 5 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/CircleMPQSolver.cpp:93]: no. i'll fix it.

File: compiler/circle-mpqsolver/src/bisection/BisectionSolver.h [hunk 11]
```diff
@@ -64,6 +64,13 @@ class BisectionSolver final : public MPQSolver
    */
   void algorithm(Algorithm algorithm);
 
+  /**
+   * @brief set path of visq_file to be used in 'auto' mode
+   * for handling which way of spliting the neural network
+   * will be the best
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/bisection/BisectionSolver.h:70]: `@brief` would be better in one line for a brief explanation.
and use `@detail` for long one.
```
@brief   set visq_file path to be used in 'auto' mode
@details this is used to handle which way (8 or 16bit) of
         splitting the neural network will be the best for accuracy.
```

Q) is like like just calling `setVisqPath` will set to auto mode?

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/BisectionSolver.h:70]: > is like like just calling `setVisqPath` will set to auto mode?

No. `setVisqPath` just assumes that `auto` will be used, but to use it `algorithm(Algorithm::Auto)` should be called explicitly.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/bisection/BisectionSolver.h]: > `@brief` would be better in one line for a brief explanation.
> and use `@details` for long one.

done

**Comment by seanshpark** [comment]
Thanks for the update. changes look good to me.
please wait review from @jinevening as he is currently on biz-trip.

**Comment by stamalakhov** [comment]
> Thanks for the update. changes look good to me. please wait review from @jinevening as he is currently on biz-trip.

ok.

File: compiler/visq/visq [hunk 12]
```diff
@@ -107,6 +123,64 @@ def _run_dalgona(model, data, analysis, save_dir):
...
+        #print("advance on data + {fp_dir} + {fq_dir}".format(fp_dir = fp32_dir, fq_dir = fq_dir))
+        _run_dalgona(models[0], data_batch_path, dump_fp32_py, fp32_dir)
+        copyfile(fp32_dir + '/tensors.txt', fq_dir + '/tensors.txt')
+        _run_dalgona(models[1], data_batch_path, dump_fq_py, fq_dir)
+
+        for metric_key in computers:
+            computers[metric_key][0].advance_on(fp32_dir, fq_dir)
+
+
+def _run_simultaneously(models, data, computers, batch_size):
```
Comment (jinevening) [inline compiler/visq/visq:160]: The name looks a bit confusing. Maybe `_run_batch` or something else would be better.

**Comment by jinevening** [comment]
Sorry for the late review. The overall change looks good to me. This PR has multiple changes, so please split them when you land it.

For example,

1. visq with advance_on
2. visq with batch
3. visq with tae
4. visq-unittest with tae
5. Use visq error in circle-mpqsolver

File: compiler/visq/visq (hunk 12 above)
Comment (stamalakhov) [reply compiler/visq/visq:160]: > The name looks a bit confusing. Maybe `_run_batch` or something else would be better.

ok

**Comment by stamalakhov** [comment]
> Sorry for the late review. The overall change looks good to me. This PR has multiple changes, so please split them when you land it.
> 
> For example,
> 
>     1. visq with advance_on
> 
>     2. visq with batch
> 
>     3. visq with tae
> 
>     4. visq-unittest with tae
> 
>     5. Use visq error in circle-mpqsolver

ok.

- [ ] visq with advance_on
- [ ] visq with batch
- [ ] visq with tae
- [ ] visq-unittest with tae
- [ ] use visq error in circle-mpqsolver

**Comment by stamalakhov** [comment]
seems like the whole draft is merged for now.

**Pull request closed without merging**